subsys/net/ip/tcp.c

```diff
@@ -2347,25 +2347,32 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
 				th_seq(th) == conn->ack)) {
 			k_work_cancel_delayable(&conn->establish_timer);
 			tcp_send_timer_cancel(conn);
-			next = TCP_ESTABLISHED;
 			tcp_conn_ref(conn);
 			net_context_set_state(conn->context,
 					      NET_CONTEXT_CONNECTED);
 
-			if (conn->accepted_conn) {
-				if (conn->accepted_conn->accept_cb) {
-					conn->accepted_conn->accept_cb(
-						conn->context,
-						&conn->accepted_conn->context->remote,
-						sizeof(struct sockaddr), 0,
-						conn->accepted_conn->context);
-				}
-
-				/* Make sure the accept_cb is only called once.
+			if (conn->accepted_conn == NULL ||
+			    conn->accepted_conn->accept_cb == NULL) {
+				/* In case of no accept_cb registered,
+				 * application will not take ownership of the
+				 * connection. To prevent connection leak, unref
+				 * the TCP context and put the connection into
+				 * active close (TCP_FIN_WAIT_1).
 				 */
-				conn->accepted_conn = NULL;
+				net_tcp_put(conn->context);
+				break;
```

Comment:
This break will skip any possible data received with the ACK. Are you deliberately ignoring this? This means that it might be retransmitted in FIN1 where it will receive a reset

Reply:
In case there is no `accept_cb`, the data would not be processed anyway, as there would be no `recv_cb` registered, so yeah, `break` here was intentional. I think that a potential RST in FIN1 state in such case may actually be a reasonable choice, as this clearly indicates that the data was **not** delivered. WDYT?

Reply:
Sounds like this ok, it would be code wist pretty nasty to send a RST from this point and go to TIME_WAIT.